waveform_editor/dependency_graph.py

```diff
@@ -0,0 +1,65 @@
+from waveform_editor.derived_waveform import DerivedWaveform
+
+
+class DependencyGraph:
+    def __init__(self):
+        self.graph = {}
+
+    def add_node(self, name, dependencies):
+        self.graph[name] = set(dependencies)
+        try:
+            self.detect_cycles()
+        except RuntimeError as e:
+            del self.graph[name]
+            raise e
+
+    def remove_node(self, name):
+        if name in self.graph:
+            del self.graph[name]
+        for deps in self.graph.values():
+            deps.discard(name)
+
+    def rename_node(self, old_name, new_name, config):
+        dependents = {node for node, deps in self.graph.items() if old_name in deps}
+        for dependent_name in dependents:
+            dependent_waveform = config[dependent_name]
+            if isinstance(dependent_waveform, DerivedWaveform):
+                dependent_waveform.rename_dependency(old_name, new_name)
+
+        if old_name in self.graph:
+            self.graph[new_name] = self.graph.pop(old_name)
+
+        for dependencies in self.graph.values():
+            if old_name in dependencies:
```

Comment:
You already determined this and stored it in the dependents list. Please reuse it :)